src/Compilers/CSharp/Test/Semantic/SourceGeneration/TestGenerators.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.IO;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.Semantic.UnitTests.SourceGeneration
+{
+    internal class SingleFileTestGenerator : ISourceGenerator
+    {
+        private readonly string _content;
+        private readonly string _hintName;
+
+        public SingleFileTestGenerator(string content, string hintName = "generatedFile")
+        {
+            _content = content;
+            _hintName = hintName;
+        }
+
+        public void Execute(SourceGeneratorContext context)
+        {
+            context.AdditionalSources.Add(this._hintName, SourceText.From(_content, Encoding.UTF8));
+        }
+
+        public void Initialize(InitializationContext context)
+        {
+        }
+    }
+
+    internal class CallbackGenerator : ISourceGenerator
+    {
+        private readonly Action<InitializationContext> _onInit;
+        private readonly Action<SourceGeneratorContext> _onExecute;
+        private readonly string _sourceOpt;
```

Comment:
Rather than take the source as a parameter to the type why not change `_onExecute` to be `Func<SourceGeneratorContext, string?>`?

Reply:
Most of the uses of this generator don't actually add source. We just increment a counter or whatever to check it was called. Making the callback return source just makes it noiser

---
In reply to: [434194957](https://github.com/dotnet/roslyn/pull/44768#discussion_r434194957) [](ancestors = 434194957)